### PR TITLE
x86, decode: fix 4-level page table walk

### PR DESCRIPTION
**Pull request by chrisguikema**
This commit fixes the page table walk when using 4-level mappings. Linux can use 64-bit memory to store tables, and masking off the upper 32-bits prevents the walk from functioning properly. Additionally, the physical instruction calculation was using the incorrect logical address bits for each level. Finally, the last PTE doesn't have the 7th bit flag a valid page, we just check the first bit to ensure the entry is present, and calculate the PTE address.

### CODE REVIEW

File: libsel4vm/src/arch/x86/processor/decode.c [hunk 1]
```diff
@@ -27,9 +27,17 @@ Author: W.A.
 /* TODO are these defined elsewhere? */
 #define IA32_PDE_SIZE(pde) (pde & BIT(7))
 #define IA32_PDE_PRESENT(pde) (pde & BIT(0))
+
+#ifdef CONFIG_X86_64_VTX_64BIT_GUESTS
```
Comment (Indanz) [inline libsel4vm/src/arch/x86/processor/decode.c:31]: Wouldn't it make more sense to always define it like this if the host is 64-bit? The only reason no to have it like that is for 32-it x86, right?

**Review by Indanz — APPROVED** [review]
Overall it seems an improvement without breaking anything.

File: libsel4vm/src/arch/x86/processor/decode.c (hunk 1 above)
Comment (chrisguikema) [reply libsel4vm/src/arch/x86/processor/decode.c:31]: I don't think so. Without `CONFIG_X86_64_VTX_64BIT_GUESTS`, with a 64-bit host you have a 32-bit guest, which cannot have 64-bit memory, so all of the page tables are in low memory where a 32-bit mask is appropriate.

**Comment by chrisguikema** [comment]
Can this be merged?

**Review by abrandnewusername — APPROVED** [review]

File: libsel4vm/src/arch/x86/processor/decode.c (hunk 1 above)
Comment (Indanz) [reply libsel4vm/src/arch/x86/processor/decode.c:31]: But a 32-bit guest can have PAE support too, right?

So not really related to your changes, but who controls what addresses `vm_fetch_instruction()` deals with? That is, are all those assumptions and asserts safe? What if `eip` points to invalid memory?

Comment (chrisguikema) [reply libsel4vm/src/arch/x86/processor/decode.c:31]: No. 32-bit guests do not have PAE.

image

There's also a check to enforce this:

```
    /* ensure that PAE is not enabled */
#ifndef CONFIG_X86_64_VTX_64BIT_GUESTS
    if (cr4 & X86_CR4_PAE) {
        ZF_LOGE("Do not support walking PAE paging structures");
        return -1;
    }
#endif /* not CONFIG_X86_64_VTX_64BIT_GUESTS */
```

Additionally, the macros deal specifically with physical addresses, which in a 32-bit guest will always be 32-bit.

**Pull request merged**